### PR TITLE
unify ports view between web and desktop

### PR DESCRIPTION
**Pull request by mustard-mh**
Relate PR https://github.com/gitpod-io/gitpod/issues/7408

## TODO
- [x] ‼️  Bring preview port view back, and add a deprecation warning
- [x] ‼️ Observe, add another field `isWebview` in event `vscode_execute_command_gitpod_ports`
- [x] ‼️ Add svelte build to build command
- [x] ⁉️ Use `@vscode/webview-ui-toolkit`
- [ ] Doc tracking event data in notion
- [x] Bring fast public/private switch back
- [x] Clean up PR - remve some useless `TODO`
- [x] ~~Use beautiful svg icons🙈~~ maybe not necessary? 
- [x] Other `TODO(hw)` in code
- [x] Flip icons of public/private and tunnel/un-tunnel
- [x] Focus on new port view if user clicks port in status bar or DEPRECATE notice in Remote Explorer

## How to Test

https://hw-vs-debug-2.preview.gitpod-dev.com/workspaces

- Make sure use `latest` code
- Start a workspace
- Check if it will focus on new port view when we click DEPRECATE msg in remote explorer
- Check if click status bar ports info will focus on new port view
- Check if new ports view works as expected and data align to our previous port view

### Example cases for port features

Open browser `Remote Explorer`(previous) and new `Ports View`(new), to see if they change together

- Make port public and private
- Open browser and preview
- Copy in new port view

Open local companion with prev env host
- See if description of port changed
- See if tunnul related icons showup
   - (It should be, but it won't, don't know why, I tested with production env latest code not worked too (stable worked)
 - See if open preview and open browser turn into `localhost:xxxx`

### CODE REVIEW

**Comment by mustard-mh** [comment]
I rebased it, could you help with those questions @jeanp413

File: extensions/gitpod-web/src/ports/index.ts [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+	}
+	updateStatusBar();
+	context.subscriptions.push(portViewProvider.onDidChangePorts(() => updateStatusBar()));
+    // TODO(hw): What's this?
+	// context.subscriptions.push(vscode.commands.registerCommand('gitpod.ports.reveal', () => {
+	// 	workspaceView.reveal(gitpodWorkspaceTreeDataProvider.ports, {
+	// 		focus: true,
+	// 		expand: true
+	// 	});
+	// }));
```
Comment (jeanp413) [inline extensions/gitpod-web/src/ports/index.ts:164]: a command to show the ports view, it's used in the walktrough if you search for `gitpod.ports.reveal`

File: extensions/gitpod-web/src/ports/portView.ts [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+    toSvelteObject() {
+        return {
+            info: this.info,
+            status: this.status,
+        };
+    }
+
+    openExternal() {
+        // TODO(hw): Why we use `localUrl` before, it's always localhost:xxx for me. Maybe it's because I am debuging?
+        return vscode.env.openExternal(vscode.Uri.parse(this.status.exposed?.url ?? this.info.localUrl));
```
Comment (jeanp413) [inline extensions/gitpod-web/src/ports/portView.ts:137]: it will be resolved later in `gitpod/worbench.ts` in `resolveExternalUri`

**Comment by akosyakov** [comment]
It seems to be quite buggy yet. The ports view was empty for me and then when I try to edit .gitpod.yml I ended up in such state:
Screenshot 2022-06-10 at 10 24 27

In regards of UI:
- tab should be called just `Ports` as in VS Code
- we can remove `Origin` for now, since we are exposing and it does not make much sense for us
- we can rename `Running Process` to `Description` and display there 
- we should preserve status messages and tooltips from the old view and find place to display it too

**Comment by akosyakov** [comment]
btw we should not commit bundled artifacts

**Comment by akosyakov** [comment]
Generally we should test all old functionality and make sure that it works with new view, all commands, all possible user paths, i.e. clikcing something somewhere, editing .gitpod.yml, serving/stopping ports and so on. It would be good if you start writing down such tests cases by analysing existing code in `How to test` section.

File: extensions/gitpod-web/src/extension.ts [hunk 3]
```diff
@@ -5,25 +5,20 @@
...
 /// <reference path='../../../src/vscode-dts/vscode.proposed.resolvers.d.ts'/>
 
-import * as workspaceInstance from '@gitpod/gitpod-protocol/lib/workspace-instance';
-import * as grpc from '@grpc/grpc-js';
 import * as fs from 'fs';
 import * as os from 'os';
 import * as uuid from 'uuid';
 import { GitpodPluginModel, GitpodExtensionContext, setupGitpodContext, registerTasks, registerIpcHookCli } from 'gitpod-shared';
 import { GetTokenRequest } from '@gitpod/supervisor-api-grpc/lib/token_pb';
-import { PortsStatus, ExposedPortInfo, PortsStatusRequest, PortsStatusResponse, PortAutoExposure, PortVisibility, OnPortExposedAction } from '@gitpod/supervisor-api-grpc/lib/status_pb';
```
Comment (akosyakov) [inline extensions/gitpod-web/src/extension.ts:15]: let's revert changes to this file

good PR should make as less a possible to this file to avoid breaking existing logic

**Comment by mustard-mh** [comment]
> btw we should not commit bundled artifacts

I add these bundle files to make `preview env` works without changing the build command, will make a clean later

> let's revert changes to this file
>
> good PR should make as less a possible to this file to avoid breaking existing logic

This large `extension.ts` file makes logic unclear for me, that's why I move it. Revert works for me since tree view should be back

File: BUILD.yaml [hunk 4]
```diff
@@ -15,3 +15,4 @@ packages:
         - ["yarn", "--cwd", "./install/build", "compile"]
         - ["yarn", "--cwd", "./install", "compile"]
         - ["yarn", "--cwd", "./install", "download-builtin-extensions"]
+        - ["yarn", "--cwd", "./install", "svelte:gitpod-web"]
```
Comment (filiptronicek) [inline BUILD.yaml:18]: Can we add this to the prepublish step of the extension build? I would say it's better to have less steps here, since this is just a part of one extension.

https://github.com/gitpod-io/openvscode-server/blob/37230ed56a61626ab916ad9285d5789db9e3e9f9/extensions/gitpod-web/package.json#L525

**Comment by akosyakov** [comment]
> This large extension.ts file makes logic unclear (and code harder to find) for me, that's why I move it to group ports related code. Revert works for me since tree view should be back

It is fine to extract ports model logic independent from presentations, but be careful to avoid regressions 🙏

File: extensions/gitpod-web/package.json [hunk 5]
```diff
@@ -1,510 +1,557 @@
 {
-  "name": "gitpod-web",
```
Comment (jeanp413) [inline extensions/gitpod-web/package.json:2]: could you revert this formatting change or put in a separate commit it creates noise to review :pray:

File: extensions/package.json [hunk 6]
```diff
@@ -9,12 +9,17 @@
...
   "scripts": {
     "postinstall": "node ./postinstall.mjs",
-    "compile": "../node_modules/.bin/tsc -b",
-    "watch": "../node_modules/.bin/tsc -b -w"
+    "compile": "yarn compile:gitpod-web-res && yarn compile:ts",
+    "watch": "npm-run-all watch:ts watch:gitpod-web-res",
+	"compile:ts": "../node_modules/.bin/tsc -b",
+	"watch:ts": "../node_modules/.bin/tsc -b -w",
+	"compile:gitpod-web-res": "yarn --cwd gitpod-web svelte:build",
+	"watch:gitpod-web-res": "yarn --cwd gitpod-web svelte:dev"
```
Comment (jeanp413) [inline extensions/package.json:17]: fix formatting

Comment (mustard-mh) [reply extensions/package.json:17]: I have no idea why the indent is not the same, even though I copy that indent from above...

File: extensions/gitpod-web/package.json (hunk 5 above)
Comment (mustard-mh) [reply extensions/gitpod-web/package.json:2]: I will revert and use **save without format** later

File: extensions/package.json (hunk 6 above)
Comment (jeanp413) [reply extensions/package.json:17]: I think you are using tabs and should be spaces

**Comment by mustard-mh** [comment]
Questions:

- ⁉️ Should we replace svgs with some outline svgs, their font-weight too big (I found them from fontawesome free icons)?
     (The warning status is a mock data here) @gtsiolis 

|Hover Options| Status Icon |
|-|-|
|image|image|

- ⁉️ A properly title for port view

|Before Click| After Click |
|-|-|
|image|image|

File: extensions/gitpod-web/src/extension.ts [hunk 7]
```diff
@@ -400,85 +511,184 @@ export class GitpodWorkspaceTreeDataProvider implements vscode.TreeDataProvider<
...
+		const ports = Array.from(this.portMap.values()).map(e => e.toSvelteObject());
+		this._view?.webview.postMessage({ command: 'updatePorts', ports });
+	}
+
+	private onHtmlCommand() {
+		this._view?.webview.onDidReceiveMessage(async (message: { command: PortCommand; port: { info: PortInfo; status: PortsStatus.AsObject } }) => {
+			if (message.command === 'queryPortData') {
+				this.updateHtml();
+				return;
+			}
```
Comment (mustard-mh) [inline extensions/gitpod-web/src/extension.ts:656]: We need this because `onDidChangeVisibility` does not work sometimes... 

That's why [Anton](https://github.com/gitpod-io/openvscode-server/pull/361#issuecomment-1152110710) needs to trigger a ports update to load this Webview first time

**Comment by filiptronicek** [comment]
> Should we replace SVGs with some outline SVGs? Their font-weight too big

@mustard-mh could we use [Codicons](https://microsoft.github.io/vscode-codicons/dist/codicon.html) for everything?

**Comment by mustard-mh** [comment]
> > Should we replace SVGs with some outline SVGs? Their font-weight too big
> 
> @mustard-mh could we use [Codicons](https://microsoft.github.io/vscode-codicons/dist/codicon.html) for everything?

👍  I will try with this package [svelte-codicons](https://www.npmjs.com/package/svelte-codicons)

**Comment by akosyakov** [comment]
You did have a look at https://github.com/microsoft/vscode-webview-ui-toolkit right? It is provided by MS to have native like experience. I am not sure about tech choices with Svelte and so on.

**Comment by mustard-mh** [comment]
> You did have a look at https://github.com/microsoft/vscode-webview-ui-toolkit right?

Never. if I know it, I will not draw a table myself 😅

**Comment by mustard-mh** [comment]
Hey @filiptronicek , I use `@vocode/codicons` instead custom SVG now, thanks for this suggestion, and thanks @jeanp413 to help me with `package.json > nohoist` 🙏

**Comment by akosyakov** [comment]
> Never. if I know it, I will not draw a table myself 😅

Does it make sense to switch? it seems to have codicons already in.

**Comment by mustard-mh** [comment]
> Does it make sense to switch? it seems to have codicons already in.

`vscode-webview-ui-toolkit` its [`Data Grid`](https://microsoft.github.io/vscode-webview-ui-toolkit/?path=/story/library-data-grid--with-custom-titles) is not fit what we want, so I will keep this custom table.

**Comment by gtsiolis** [comment]
Thanks for the ping in https://github.com/gitpod-io/openvscode-server/pull/361#issuecomment-1156102513, @mustard-mh! 🏓 

1️⃣ Do you still need any UX input on this? If yes, how can I see the changes in action? I'm still seeing the same icons on VS Code Browser, VS Code Desktop.

| VS Code (Browser) | VS Code (Desktop) |
|-|-|
| Screenshot 2022-06-16 at 3 24 49 PM | Screenshot 2022-06-16 at 3 26 25 PM |

2️⃣ FWIW, going with the VS Code font icon (Codeicons[[1](https://github.com/microsoft/vscode-codicons)]) as @filiptronicek suggested in https://github.com/gitpod-io/openvscode-server/pull/361#issuecomment-1156117943 sounds good and much more preferable. Alternatively, if we ever need any custom SVG icon, let's go with Heroicons[[2](https://heroicons.com/)], which is what we've been using for the dashboard. 💡

**Comment by akosyakov** [comment]
> vscode-webview-ui-toolkit its [Data Grid](https://microsoft.github.io/vscode-webview-ui-toolkit/?path=/story/library-data-grid--with-custom-titles) is not quite fit what we want, so I will keep this custom table.

ok the make sure that all selection colors are applied properly, right now it looks alient compare to native views 🙏

**Comment by mustard-mh** [comment]
> ok the make sure that all selection colors are applied properly, right now it looks alient compare to native views 🙏

I build a new branch to use `vscode-webview-ui-toolkit`, you can try it here https://werft.gitpod-dev.com/job/gitpod-build-hw-vs-debug-2.0 , will update this PR later

File: extensions/gitpod-web/src/extension.ts [hunk 8]
```diff
@@ -399,86 +506,181 @@ export class GitpodWorkspaceTreeDataProvider implements vscode.TreeDataProvider<
...
-				if (port.tunnel) {
-					port.contextValue = (isPortTunnelPublic ? 'network-' : 'host-') + port.contextValue;
-				}
-				if (!accessible && portStatus.getAutoExposure() === PortAutoExposure.FAILED) {
-					port.contextValue = 'failed-' + port.contextValue;
-				}
-				if (isExposedServedGitpodWorkspacePort(port) && !isExposedServedPort(prevStatus)) {
-					this.onDidExposeServedPortEmitter.fire(port);
-				}
+				// if (isExposedServedGitpodWorkspacePort(port.port) && !isExposedServedPort(prevStatus)) {
```
Comment (akosyakov) [inline extensions/gitpod-web/src/extension.ts:524]: why is it commented out. did we move it somewhere else or some logic was changed?

File: extensions/gitpod-web/src/extension.ts [hunk 9]
```diff
@@ -399,86 +506,181 @@ export class GitpodWorkspaceTreeDataProvider implements vscode.TreeDataProvider<
 			const toClean = new Set<number>(this.ports.ports.keys());
 			const portsList = this.portStatus.getPortsList();
 			for (const portStatus of portsList) {
-				const currentStatus = portStatus.toObject();
-				const { name, localPort, description, exposed, served } = currentStatus;
+				const localPort = portStatus.getLocalPort();
+				const tunnel = this.tunnelMap.get(localPort);
```
Comment (akosyakov) [inline extensions/gitpod-web/src/extension.ts:510]: it is something new, why do we do it?

Comment (akosyakov) [reply extensions/gitpod-web/src/extension.ts:510]: is it because we have 2 duplicated structures, so we have 2 GitpodWorkspacePort for 2 ports? Why don't introduce PortsModel decoupled from presentations?

File: extensions/gitpod-web/webview-ui/src/components/HoverOptions.svelte [hunk 10]
```diff
@@ -0,0 +1,82 @@
+<script lang="ts">
```
Comment (akosyakov) [inline extensions/gitpod-web/webview-ui/src/components/HoverOptions.svelte:1]: Could you add Gitpod copyright headers to all new files please?

File: extensions/gitpod-web/package.json [hunk 11]
```diff
@@ -417,6 +417,16 @@
           "id": "gitpod.workspace",
           "name": "Gitpod Workspace"
         }
+      ],
+      "portView": [
+        {
+          "id": "gitpod.portView",
+          "name": "",
```
Comment (akosyakov) [inline extensions/gitpod-web/package.json:424]: name, contextualTiel and icon are left out blank on purpose?

File: extensions/gitpod-web/src/extension.ts (hunk 9 above)
Comment (mustard-mh) [reply extensions/gitpod-web/src/extension.ts:510]: `GitpodWorkspacePort` is a `PortModel` now, only provide data that `treeView` and `webview` need

File: extensions/gitpod-web/src/extension.ts (hunk 8 above)
Comment (mustard-mh) [reply extensions/gitpod-web/src/extension.ts:524]: Yeah, we will handle this only for one place, any of them is ok since we will not DEPRECATE treeView

File: extensions/gitpod-web/package.json [hunk 12]
```diff
@@ -516,25 +516,30 @@
...
-    "build": "yarn build:webview && yarn compile:ts",
+    "build": "yarn jest; yarn build:webview && yarn compile:ts",
     "watch": "npm-run-all -lp watch:webview watch:ts",
     "compile:ts": "tsc -p ./",
     "watch:ts": "tsc -watch -p ./",
     "install:webview": "yarn --cwd ./webview-ui --modules-folder node_modules",
     "watch:webview": "yarn --cwd ./webview-ui run dev",
-    "build:webview": "yarn --cwd ./webview-ui run build"
+    "build:webview": "yarn --cwd ./webview-ui run build",
+    "test": "jest"
```
Comment (jeanp413) [inline extensions/gitpod-web/package.json:526]: we should use mocha as vscode does, no need for a new test framework to learn :smile:

File: extensions/gitpod-web/webview-ui/package.json [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+		"rollup-plugin-livereload": "^2.0.0",
+		"rollup-plugin-svelte": "^7.0.0",
+		"rollup-plugin-terser": "^7.0.0",
+		"svelte": "^3.0.0",
+		"svelte-check": "^2.0.0",
+		"svelte-preprocess": "^4.0.0",
+		"tslib": "^2.0.0",
+		"typescript": "^4.0.0"
+	}
+}
```
Comment (jeanp413) [inline extensions/gitpod-web/webview-ui/package.json:35]: Could we just use one package.json? I see other extensions that also have webviews and only have one `package.json`

File: extensions/gitpod-web/package.json (hunk 12 above)
Comment (mustard-mh) [reply extensions/gitpod-web/package.json:526]: I tried to setup with mocha, but failed 🙈

File: extensions/gitpod-web/webview-ui/package.json (hunk 13 above)
Comment (mustard-mh) [reply extensions/gitpod-web/webview-ui/package.json:35]: No, this file structure is learn from an official demo. It makes things (both structure and code review) easier than before, I'll keep it.

Comment (jeanp413) [reply extensions/gitpod-web/webview-ui/package.json:35]: File structure is fine, I meant to just move `package.json` content i.e `scripts`, `dependencies`, `devDependencies` to the parent package.json so there's only one `yarn.lock` file for gitpod-web extension

Comment (mustard-mh) [reply extensions/gitpod-web/webview-ui/package.json:35]: We need dependencies like `@vscode/codicons` `@vscode/webview-ui-toolkit` to install in `current folder node_moddules`. with `extensions/` folder's setup, some of them will be inside `extensions/node_modules` this is very bad for the build, and I don't want this to happen again (configure of `nohoist` sometimes make things complex).

And these PKGs will truly work on `webview-ui` only, after `webview-ui` build success, we don't need them anymore.

Everything related to webview is only existed in `webview-ui` folder, I think that's good, it only works to build what extension needs

Comment (mustard-mh) [reply extensions/gitpod-web/webview-ui/package.json:35]: I don't know if it's worth moving `scripts/dep/dev-dep` out, since other extensions with webview don't use front-framework

File: extensions/gitpod-web/package.json (hunk 12 above)
Comment (mustard-mh) [reply extensions/gitpod-web/package.json:526]: I'll have another try with `mocha` setup

File: extensions/gitpod-web/package.json [hunk 14]
```diff
@@ -484,23 +495,51 @@
           "type": "boolean",
           "description": "Control whether to prompt to open in VS Code Desktop on page load.",
           "default": false
+        },
+        "gitpod.experiment.enable": {
```
Comment (jeanp413) [inline extensions/gitpod-web/package.json:499]: Change to `gitpod.experiment.portsWebview`

File: extensions/gitpod-web/src/extension.ts [hunk 15]
```diff
@@ -399,87 +356,191 @@ export class GitpodWorkspaceTreeDataProvider implements vscode.TreeDataProvider<
...
+
 export function registerPorts(context: GitpodExtensionContext): void {
-	const gitpodWorkspaceTreeDataProvider = new GitpodWorkspaceTreeDataProvider(context);
-	const workspaceView = vscode.window.createTreeView('gitpod.workspace', {
-		treeDataProvider: gitpodWorkspaceTreeDataProvider,
-	});
-	context.subscriptions.push(workspaceView);
+
+	const experimentCfg = vscode.workspace.getConfiguration('gitpod.experiment');
+	const isExperimentMode = vscode.version.includes('insider') || experimentCfg.get<boolean>('enable') === true;
```
Comment (jeanp413) [inline extensions/gitpod-web/src/extension.ts:527]: Don't check for insider, only user config flag `gitpod.experimental.portsWebview.enabled`

File: extensions/gitpod-web/webview-ui/public/index.html [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+<head>
+	<title>Hello World</title>
+	<meta charset='utf-8'>
+	<meta name='viewport' content='width=device-width,initial-scale=1'>
+	<link rel='stylesheet' href='/build/bundle.css'>
+	<script defer src='/build/bundle.js'></script>
+</head>
+<body>
+</body>
+</html>
```
Comment (jeanp413) [inline extensions/gitpod-web/webview-ui/public/index.html:12]: Delete this?

File: extensions/gitpod-web/webview-ui/package.json (hunk 13 above)
Comment (jeanp413) [reply extensions/gitpod-web/webview-ui/package.json:35]: What is `scripts/dep/dev-dep`?

**Comment by jeanp413** [comment]
Usually `types` folder is used to put declaration files i.e. `*.d.ts` so maybe rename it to something else.
Also rename `webview-ui` folder to something more specific like `portsView`

File: extensions/gitpod-web/webview-ui/package.json (hunk 13 above)
Comment (mustard-mh) [reply extensions/gitpod-web/webview-ui/package.json:35]: scripts, dependencies, devDependencies

File: extensions/gitpod-web/src/extension.ts (hunk 15 above)
Comment (mustard-mh) [reply extensions/gitpod-web/src/extension.ts:527]: Speak with Anton that insiders will have this port view, I add this config to make stable can enable it too

File: extensions/gitpod-web/webview-ui/public/index.html (hunk 16 above)
Comment (mustard-mh) [reply extensions/gitpod-web/webview-ui/public/index.html:12]: We can debug with webview-ui only (not run install built-in extensions) with this `index.html` file

File: extensions/gitpod-web/src/extension.ts (hunk 15 above)
Comment (jeanp413) [reply extensions/gitpod-web/src/extension.ts:527]: I was thinking by default should `true` in `gp-code/main`, but when we create the release branch we set it to `false`, no need for additional `insider` check
cc @akosyakov

Comment (mustard-mh) [reply extensions/gitpod-web/src/extension.ts:527]: If so, we need to add this step (turn default `true` into `false) into release steps

**Comment by mustard-mh** [comment]
> Usually types folder is used to put declaration files i.e. *.d.ts so maybe rename it to something else.

I tried to put them all into something like `global.d.ts` there must have some reasons why I use `.ts` (), I'll try it again

> Also rename webview-ui folder to something more specific like portsView

Not sure if we will only have port view to use webview, good for me to do it too since we only have one for now

cc @jeanp413

File: extensions/gitpod-web/webview-ui/public/index.html (hunk 16 above)
Comment (jeanp413) [reply extensions/gitpod-web/webview-ui/public/index.html:12]: mmm is it really useful? I ask that because vscode will inject the API and css variables, so if you run it with this html those won't be defined

**Comment by jeanp413** [comment]
> I tried to put them all into something like global.d.ts there must have some reasons why I use .ts (), I'll try it again

Using `.ts` here is correct, `*.d.ts.` is only needed if you are importing some library without types or you are writing a library 
exposing some API

File: extensions/gitpod-web/webview-ui/public/index.html (hunk 16 above)
Comment (mustard-mh) [reply extensions/gitpod-web/webview-ui/public/index.html:12]: 😂 Yes, I'm not using this too, I create a extension locally to debug webview

File: extensions/gitpod-web/src/extension.ts (hunk 15 above)
Comment (mustard-mh) [reply extensions/gitpod-web/src/extension.ts:527]: > I was thinking by default should true in gp-code/main, but when we create the release branch we set it to false, no need for additional insider check

I'll go with this if we can make changes in insider release. cc @jeanp413

File: extensions/gitpod-web/package.json (hunk 12 above)
Comment (mustard-mh) [reply extensions/gitpod-web/package.json:526]: @jeanp413  If we use `mocha`, 

- We need `ts-node/register` for `.ts` testing
- `ts-node` must install in current folder, otherwise, mocha will said `ts-node/register` not found
- After `nohoist` ts-node in current folder, mocha will say `gitpod-shared` not found

We can't move gitpod related module inside extension folder itself, otherwise their refer of `node_modules` will be different, and will cause an error like before [internal chat](https://gitpod.slack.com/archives/C01KGM9BH54/p1655831041154569?thread_ts=1655827615.556309&cid=C01KGM9BH54)

Maybe there have another ways, but I can't found it

File: extensions/gitpod-web/webview-ui/package.json (hunk 13 above)
Comment (jeanp413) [reply extensions/gitpod-web/webview-ui/package.json]: I added `nohoist` back excluding the gitpod, grpc packages that caused the previous issue, could you rebase and test it :pray:

File: extensions/gitpod-web/package.json (hunk 12 above)
Comment (jeanp413) [reply extensions/gitpod-web/package.json:526]: why do we need ts-node? `.test.ts` files should be compiled to js too

Comment (mustard-mh) [reply extensions/gitpod-web/package.json:526]: At least for now we decided to use `jest`, it'll inside our build steps (yes, werft failed again 😅 will try clean slate later)

|Test in Werft Build| 
|-|
|image|

| Test Coverage |
|-|
|SCR-20220624-vun|

File: extensions/gitpod-web/src/util/port.ts [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+/// <reference path='../../../../src/vscode-dts/vscode.d.ts'/>
+/// <reference path='../../../../src/vscode-dts/vscode.proposed.resolvers.d.ts'/>
+
+import * as workspaceInstance from '@gitpod/gitpod-protocol/lib/workspace-instance';
+import { GitpodExtensionContext } from 'gitpod-shared';
+import { PortsStatus, PortAutoExposure, PortVisibility } from '@gitpod/supervisor-api-grpc/lib/status_pb';
+import { TunnelVisiblity, TunnelPortRequest } from '@gitpod/supervisor-api-grpc/lib/port_pb';
+import { URL } from 'url';
+import * as util from 'util';
+import * as vscode from 'vscode';
```
Comment (jeanp413) [inline extensions/gitpod-web/src/util/port.ts:14]: @mustard-mh I gave a little more thought about this, let's decouple `ports.ts` from vscode API so you can remove the vscode API mocks, this will also ensure that if someone wants to add more tests in the future they don't have the wrong idea of keep mocking vscode API, if we really need vscode API in our tests in the future then we'll have to look into moving gitpod-web to it's own repo

File: extensions/gitpod-web/src/util/port.spec.ts [hunk 18]
```diff
@@ -0,0 +1,265 @@
+import { ExposedPortInfo, OnPortExposedAction, PortAutoExposure, PortsStatus, PortVisibility, TunneledPortInfo } from '@gitpod/supervisor-api-grpc/lib/status_pb';
```
Comment (jeanp413) [inline extensions/gitpod-web/src/util/port.spec.ts:1]: Let's sue the same filename format as other vscode tests rename to `port.test.ts` and move it inside a `test` folder

**Review by jeanp413 — APPROVED** [review]
Will keep iterating/fixing bugs in gp-code/main branch, so merging this
